cpp/src/community/leiden_impl.cuh

```diff
@@ -42,17 +45,508 @@ void check_clustering(graph_view_t<vertex_t, edge_t, false, multi_gpu> const& gr
+          shuffle_ext_vertex_value_pairs_to_local_gpu_by_vertex_partitioning(
+            handle, std::move(cluster_keys), std::move(cluster_weights));
+      }
+
+      first_iteration = false;
+    } else {
+      rmm::device_uvector<weight_t> tmp_weights_buffer(vertex_weights.size(),
+                                                       handle.get_stream());  // #C
+
+      raft::copy(dendrogram->current_level_begin(),
+                 louvain_of_refined_partition.begin(),
+                 louvain_of_refined_partition.size(),
+                 handle.get_stream());
+
+      raft::copy(tmp_weights_buffer.begin(),
+                 vertex_weights.begin(),
+                 vertex_weights.size(),
+                 handle.get_stream());
+
+      thrust::sort_by_key(handle.get_thrust_policy(),
+                          louvain_of_refined_partition.begin(),
+                          louvain_of_refined_partition.end(),
+                          tmp_weights_buffer.begin());
+
+      auto pair_of_iterators_end = thrust::reduce_by_key(handle.get_thrust_policy(),
+                                                         louvain_of_refined_partition.begin(),
+                                                         louvain_of_refined_partition.end(),
+                                                         tmp_weights_buffer.begin(),
+                                                         cluster_keys.begin(),
+                                                         cluster_weights.begin());
+
+      louvain_of_refined_partition.resize(0, handle.get_stream());
+      louvain_of_refined_partition.shrink_to_fit(handle.get_stream());
+
+      tmp_weights_buffer.resize(0, handle.get_stream());
+      tmp_weights_buffer.shrink_to_fit(handle.get_stream());
+
+      cluster_keys.resize(
+        static_cast<size_t>(thrust::distance(cluster_keys.begin(), pair_of_iterators_end.first)),
+        handle.get_stream());
+      cluster_weights.resize(static_cast<size_t>(thrust::distance(cluster_weights.begin(),
+                                                                  pair_of_iterators_end.second)),
+                             handle.get_stream());
+
+      cluster_keys.shrink_to_fit(handle.get_stream());
+      cluster_weights.shrink_to_fit(handle.get_stream());
+
+      if constexpr (graph_view_t::is_multi_gpu) {
+        rmm::device_uvector<vertex_t> tmp_keys_buffer(0, handle.get_stream());  // #C
+
+        std::tie(tmp_keys_buffer, tmp_weights_buffer) =
+          shuffle_ext_vertex_value_pairs_to_local_gpu_by_vertex_partitioning(
+            handle, std::move(cluster_keys), std::move(cluster_weights));
+
+        thrust::sort_by_key(handle.get_thrust_policy(),
+                            tmp_keys_buffer.begin(),
+                            tmp_keys_buffer.end(),
+                            tmp_weights_buffer.begin());
+
+        cluster_keys.resize(tmp_keys_buffer.size(), handle.get_stream());
```

Comment:
See the code here https://github.com/rapidsai/cugraph/blob/branch-23.04/cpp/src/structure/renumber_edgelist_impl.cuh#L449 (line 449-461).

You can avoid addition resize and shrink_to_fit function calls (This can be more costly than counting # uniques first and also use more memory).